lib/validator.js

```diff
@@ -88,220 +88,258 @@ Validator.prototype.validate = function(obj, schema) {
+
+	sourceCode.push(`propertyPath = (path !== undefined ? path + ".${name}" : "${name}");`);
+	if (Array.isArray(property.compiledType)) {
+	    sourceCode.push(`res = this.checkSchemaType(${propertyValueExpr}, properties[${i}].compiledType, propertyPath, value);`);
+	} else {
+	    sourceCode.push(`res = this.checkSchemaRule(${propertyValueExpr}, properties[${i}].compiledType, propertyPath, value);`);
 	}
+	sourceCode.push("if (res !== true) {");
+	sourceCode.push(`\tthis.handleResult(errors, propertyPath, res, properties[${i}].compiledType.messages);`);
+	sourceCode.push("}");
+    }
 
-	sourceCode.push("return errors.length === 0 ? true : errors;");
+    sourceCode.push("return errors.length === 0 ? true : errors;");
 
-	compiledObject.compiledObjectFunction = new Function("value", "properties", "path", "parent", sourceCode.join("\n"));
+    compiledObject.compiledObjectFunction = new Function("value", "properties", "path", "parent", sourceCode.join("\n"));
 
-	return compiledObject;
+    return compiledObject;
 };
 
 Validator.prototype.compileSchemaType = function(schemaType) {
-	if (Array.isArray(schemaType)) {
-		// Multiple rules, flatten to array of compiled SchemaRule
-		const rules = flatten(schemaType.map(r => this.compileSchemaType(r)));
-		if (rules.length == 1) {
-			return rules[0];
-		}
-
-		return rules;
-	}
 
-	return this.compileSchemaRule(schemaType);
-};
+    if (Array.isArray(schemaType)) {
+	
+	// Multiple rules, flatten to array of compiled SchemaRule
+	const rules = flatten(schemaType.map(r => this.compileSchemaType(r)));
 
-Validator.prototype.compileSchemaRule = function(schemaRule) {
-	if (typeof schemaRule === "string") {
-		schemaRule = {
-			type: schemaRule
-		};
+	if (rules.length == 1) {
+	    return rules[0];
 	}
+	
+	return rules;
+    }
+    
+    return this.compileSchemaRule(schemaType);
+    
+};
 
-	const ruleFunction = this.rules[schemaRule.type];
-	if (!ruleFunction) {
-		throw new Error("Invalid '" + schemaRule.type + "' type in validator schema!");
-	}
+Validator.prototype.compileMessages = function(schemaType) {
 
-	let dataParameter = null;
-	let dataFunction = null;
+    let messages = {};
+    
+    if (schemaType.messages && typeof schemaType.messages == "object")
+	messages = schemaType.messages;
 
-	if (schemaRule.type === "object" && schemaRule.props) {
-		dataParameter = this.compileSchemaObject(schemaRule.props);
-		dataFunction = this.checkSchemaObject;
-	} else if (schemaRule.type === "array" && schemaRule.items) {
-		dataParameter = this.compileSchemaType(schemaRule.items);
-		dataFunction = this.checkSchemaArray;
-	}
+    if (!messages["required"])
+	messages["required"] = this.messages["required"];
+    
+    if (!messages[schemaType.type])
+	messages[schemaType.type] = this.messages[schemaType.type];
+    
+    Object.keys(schemaType).forEach(rule => {
+	if (rule == "type" || rule == "items" || rule == "props" || rule == "messages") return;
```

Comment:
Could you explain what is this code?

Reply:
In order to remove check from run time we have to prepare all possible message types at compile time. These include:

‘required’
‘string’ | ‘number’ | ‘array’ etc (i.e type)

Then we analyse the additional rules to see what other messages need to be added.

Reply:
I've just changed the if statement to use `key in obj` style rather than truthy test